src/test/java/ch/naviqore/raptor/RaptorTest.java

```diff
@@ -78,14 +54,91 @@ void routeBetweenTwoStopsOnSameRoute(RaptorTestBuilder builder) {
+            List<Connection> connections = raptor.routeEarliestArrival(sourceStops, targetStops);
+            assertEquals(1, connections.size());
+            Helpers.assertConnection(connections.getFirst(), "B", "H", departureTime, 0, 0, 1);
+        }
+
+        @Test
+        void routeFromTwoSourceStopsWithLaterDepartureTimeOnCloserStop(RaptorTestBuilder builder) {
+            Raptor raptor = builder.buildWithDefaults();
+
+            int departureTime = 8 * RaptorTestBuilder.SECONDS_IN_HOUR;
+            Map<String, Integer> sourceStops = Map.of("A", departureTime, "B",
+                    departureTime + RaptorTestBuilder.SECONDS_IN_HOUR);
+            Map<String, Integer> targetStops = Map.of("H", 0);
+
+            // fastest and only connection should be B -> H
+            List<Connection> connections = raptor.routeEarliestArrival(sourceStops, targetStops);
+            assertEquals(2, connections.size());
+            Helpers.assertConnection(connections.getFirst(), "B", "H",
+                    departureTime + RaptorTestBuilder.SECONDS_IN_HOUR, 0, 0, 1);
+            Helpers.assertConnection(connections.get(1), "A", "H", departureTime, 1, 0, 2);
+        }
+
+        @Test
+        void routeFromStopToTwoTargetStopsNoWalkTimeToTarget(RaptorTestBuilder builder) {
+            Raptor raptor = builder.buildWithDefaults();
+
+            int departureTime = 8 * RaptorTestBuilder.SECONDS_IN_HOUR;
+            Map<String, Integer> sourceStops = Map.of("A", departureTime);
+            Map<String, Integer> targetStops = Map.of("F", 0, "S", 0);
+
+            // fastest and only connection should be A -> F
+            List<Connection> connections = raptor.routeEarliestArrival(sourceStops, targetStops);
+            assertEquals(1, connections.size());
+            Helpers.assertConnection(connections.getFirst(), "A", "F", departureTime, 0, 0, 1);
+        }
+
+        @Test
+        void routeFromStopToTwoTargetStopsWithWalkTimeToTarget(RaptorTestBuilder builder) {
+            Raptor raptor = builder.buildWithDefaults();
+
+            int departureTime = 8 * RaptorTestBuilder.SECONDS_IN_HOUR;
+            Map<String, Integer> sourceStops = Map.of("A", departureTime);
+            Map<String, Integer> targetStops = Map.of("F", RaptorTestBuilder.SECONDS_IN_HOUR, "S", 0);
+
+            // since F is closer to A than S, the fastest connection should be A -> F, but because of the hour
+            // walk time to target, the connection A -> S should be faster (no additional walk time)
+            List<Connection> connections = raptor.routeEarliestArrival(sourceStops, targetStops);
+            assertEquals(2, connections.size());
+            Helpers.assertConnection(connections.getFirst(), "A", "F", departureTime, 0, 0, 1);
+            Helpers.assertConnection(connections.get(1), "A", "S", departureTime, 1, 0, 2);
+
+            // Note since the required walk time to target is not added as a leg, the solutions will not be pareto
```

Comment:
This is a general issue with the basic raptor. The raptor does not minimize travel time but earliest arrival from one time point. To make it more suitable for travel time minimization we need to implement the range raptor and/or the multiconstraint raptor.
Also to resolve the leg creation you are right that we would need to move everything down a level into raptor. So that raptor would allow location/location routing. But this would require a spatial index of all stops in the raptor class.